crates/sui-core/src/authority.rs

```diff
@@ -225,7 +227,12 @@ pub static METRICS: Lazy<AuthorityMetrics> = Lazy::new(AuthorityMetrics::new);
 pub type StableSyncAuthoritySigner =
     Pin<Arc<dyn signature::Signer<AuthoritySignature> + Send + Sync>>;
 
-pub struct AuthorityState {
+/// Default store choices for AuthorityState to keep types simple for now
+pub type AuthorityState = TypedAuthorityState<SqlEventStore>;
+
+const DEFAULT_QUERY_LIMIT: usize = 1000;
+
+pub struct TypedAuthorityState<ES: EventStore> {
```

Comment:
Oh boy, I really don't want to see us go this route. In 3 months we will have:

`pub struct TypedAuthorityState<ES: EventStore, CS: CertificateStore, IS: IndexStore, CkS: CheckpointStore>`

If we can't use dynamic dispatch for this (why? @bmwill can you explain your concerns better? please be concrete) then I would like to see if we can think of any other option. Some ideas:

* config directives and duck typing
* enums + manual dynamic dispatch
* message passing interfaces a la LockService

Reply:
@mystenmark I don't see any comments from @bmwill .... 

I don't really think adding more types is that bad, it in fact makes each type in AuthorityState very very clear.  I do agree it looks worse ergonomically.

I really don't want to do boxed trait objects though.  Although dynamic dispatch for something like this is not likely to be a significant performance thing, but I've been taught to hate dynamic dispatch for performance work as much as possible.

There's a good solution I've used before, which is `enum_dispatch` ... it automates using enums to dispatch stuff, and it is not dynamic, since the code can in fact be inlined.    I'll explore this avenue.

Reply:
So I'm kind of stuck right now.

`enum_dispatch` is not working, and `Box<dyn EventStore>` also doesn't work, because I put in a return type that is an Associated Type, `EventIt`...  this was done intentionally so that the return type can be different type of iterator, for efficiency.  

I'd like to be able to keep this flexible return type, but it would force us to use generics.  I could return Vec<> but I feel that an iterator is much more flexible than a Vec<_> which uses a fixed amount of memory.

Reply:
Ok, I took a look at the code - I think we should just return `Vec<_>` for now. We already do that for all the existing query methods (e.g. `get_transactions_in_range`). My thoughts are:
- We don't need to solve this problem today, and when we do solve it lets come up with something that works for all queries, not just events.
- At the point at which returning a Vec becomes a problem, we will also probably want to have a streaming RPC for clients to call. And so at that point returning a stream (like you mentioned the other day) will probably be a very natural solution.
- But, maybe things will turn out differently for reasons we don't yet foresee - all the more reason to go with the dumb simple approach right now.

Reply:
I was going to switch most calls to Vec for now, but come up with something for one of them.   Just as an experiment. You'll see it soon!

Reply:
My main question is "what's the rationale for needing to be generic over these store types?"

I can agree that going super generic can have some interesting effects on the type signatures of these types making them more unwieldy

Reply:
@bmwill the initial thinking is because one is only going to have one specific implementation of each type of store at a time, so it makes sense from the type perspective that "a Sui datastore" is "a DataStore with a SQL event store" for example, and that subtype helps define what a store is.  Because the type of store should be static, not dynamic.... I wasn't thinking so much about extra signatures but just from a type standpoint what made sense, and it should definitely be static and not dynamic.